{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import skbio\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "workdir = \"/Volumes/McMinds/git_repos/gcmp_stan_symportal_2019/\"\n",
    "outdir = workdir + \"output/symbio_phylo/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A 247\n",
      "B 76\n",
      "C 706\n",
      "F 55\n",
      "D 119\n",
      "E 1\n",
      "G 28\n",
      "H 9\n",
      "I 4\n"
     ]
    }
   ],
   "source": [
    "input_file = \"all_seqs.fasta\"\n",
    "#input_file = \"72_DBV_20190609_2019-06-10_01-30-10.537772.seqs.fasta\"\n",
    "discard_unnamed = True\n",
    "clades = {}\n",
    "for seq in SeqIO.parse(workdir + \"raw_data/symbio_phylo/\" + input_file, \"fasta\"):\n",
    "        \n",
    "        if \"_\" in seq.id:\n",
    "            if discard_unnamed:\n",
    "                continue\n",
    "            else:\n",
    "                clade = seq.id.split(\"_\")[1]\n",
    "        else:\n",
    "            clade = seq.id[0]\n",
    "            \n",
    "        if clade not in clades.keys():\n",
    "            clades[clade] = []\n",
    "            \n",
    "        clades[clade].append(seq)   \n",
    "        \n",
    "clade_sizes = {}\n",
    "for x in clades.keys():\n",
    "    clade_sizes[x] = len(clades[x])\n",
    "    print(x + \" \" + str(clade_sizes[x]))\n",
    "    \n",
    "clades_sorted = sorted(clade_sizes, key=clade_sizes.get)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in clades_sorted:\n",
    "    prefix = outdir + \"its2_clade_\" + x\n",
    "    nseqs = len(clades[x])\n",
    "\n",
    "    print('Writing ' + str(nseqs) + ' clade ' + x + ' seq(s) to a file\\n')\n",
    "    SeqIO.write(clades[x], prefix + \".fasta\", \"fasta-2line\")\n",
    "    \n",
    "    if nseqs > 1:\n",
    "        print('Aligning clade ' + x + ':')\n",
    "        !mafft --thread 4 --maxiterate 1000 --ep 0 --genafpair {prefix + \".fasta\"} > {prefix + \"_aligned.fasta\"}\n",
    "        \n",
    "        #print('Building phylogeny for clade ' + x + ':')\n",
    "        #!iqtree -s {prefix + \"_aligned.fasta\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seeds = \"\"\n",
    "addseqs = \"/dev/null\"\n",
    "inds = []\n",
    "for x in clades_sorted:\n",
    "    if len(clades[x]) > 1:\n",
    "        seeds = seeds + \"--seed \" + outdir + \"its2_clade_\" + x + \"_aligned.fasta \"\n",
    "    else:\n",
    "        inds.append(clades[x])\n",
    "\n",
    "if len(inds) > 0:\n",
    "    if len(inds) == 1:\n",
    "        inds = inds[0]\n",
    "    print('Writing isolate clades to a file\\n')\n",
    "    addseqs = outdir + \"its2_concat_isolates.fasta\"\n",
    "    SeqIO.write(inds, addseqs, \"fasta-2line\")       \n",
    "        \n",
    "!mafft --thread 4 --maxiterate 1000 --ep 0 --genafpair {seeds}{addseqs} > {outdir + \"its2_all_clades_aligned.fasta\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "partitions_file = outdir + \"iqtree_partitions.nex\"\n",
    "f = open(partitions_file, \"w+\")\n",
    "\n",
    "part = 1\n",
    "f.write(\"#nexus\\n\")\n",
    "f.write(\"begin sets;\\n\")\n",
    "f.write(\"    charset ITS2 = \" + outdir + \"its2_all_clades_aligned_clean.fasta:*;\\n\")\n",
    "for gene in (\"28S\",\"23S\",\"cob\",\"coi\",\"elf\",\"psba\"):\n",
    "    part += 1\n",
    "    f.write(\"    charset \" + gene + \" = \" + workdir + \"raw_data/symbio_phylo/\" + gene + \"/forindividual tree \" + gene + \".fas:*;\\n\")\n",
    "f.write(\"end;\\n\")\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!sed 's/_seed_//g' {outdir + \"its2_all_clades_aligned.fasta\"} > {outdir + \"its2_all_clades_aligned_clean.fasta\"}\n",
    "!iqtree -nt AUTO -ntmax 4 -pre {outdir + \"allgenes\"} -mrate E,I,G,I+G,R,H,I+H -mfreq F,FO -spp {partitions_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "abundance = workdir + \"raw_data/symbio_phylo/72_DBV_20190609_2019-06-10_01-30-10.537772.profiles.absolute.txt\"\n",
    "df = pd.read_csv(abundance, sep=\"\\t\", index_col=0) \n",
    "df.drop(columns=[df.columns[0]], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "profs = dict(df.loc['ITS2 type profile'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: /Volumes/McMinds/git_repos/gcmp_stan_symportal_2019/output/symbio_phylo/its2tree/: File exists\r\n"
     ]
    }
   ],
   "source": [
    "from ete3 import Tree\n",
    "tree = Tree(outdir + \"allgenes.treefile\")\n",
    "\n",
    "alldivs = set()\n",
    "allnames = {}\n",
    "profile_map_file = outdir + \"profile_map.txt\"\n",
    "\n",
    "f = open(profile_map_file, \"w+\")\n",
    "\n",
    "for prof in profs.keys():\n",
    "    f.write(prof + \"_*\" + \" \" + prof + '\\n')\n",
    "    divs = profs[prof].replace('/',',').replace('-',',')\n",
    "    for div in divs.split(','):\n",
    "        newname = prof + \"_\" + div\n",
    "        alldivs.add(newname)\n",
    "        if div not in allnames.keys():\n",
    "            allnames[div] = [newname]\n",
    "            tree.search_nodes(name=div)[0].name = newname\n",
    "        else:\n",
    "            node = tree.search_nodes(name=allnames[div][0])[0]\n",
    "            if node.up.name == div:\n",
    "                node.up.add_child(name=newname, dist=0)\n",
    "            else:\n",
    "                node.name = div\n",
    "                node.add_child(name=allnames[div][0], dist=0)\n",
    "                node.add_child(name=newname, dist=0)\n",
    "            allnames[div].append(newname)\n",
    "f.close()\n",
    "\n",
    "tree.prune(alldivs, preserve_branch_length=True)\n",
    "gene_trees_folder = outdir + \"its2tree/\"\n",
    "pruned_tree = gene_trees_folder + \"allgenes.pruned.treefile\"\n",
    "!mkdir {gene_trees_folder}\n",
    "tree.write(format=5, outfile=pruned_tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*********************************************************\n",
      "*                                                       *\n",
      "*      STAG: Species Tree inference from All Genes      *\n",
      "*                                                       *\n",
      "*********************************************************\n",
      "\n",
      "274 species in mapping file:\n",
      "34124\n",
      "34127\n",
      "34126\n",
      "34121\n",
      "34418\n",
      "34123\n",
      "34122\n",
      "34486\n",
      "34414\n",
      "34251\n",
      "34416\n",
      "34129\n",
      "34410\n",
      "34659\n",
      "34494\n",
      "34398\n",
      "34712\n",
      "34559\n",
      "34560\n",
      "34393\n",
      "34390\n",
      "34557\n",
      "34120\n",
      "34091\n",
      "34651\n",
      "34652\n",
      "34524\n",
      "34301\n",
      "34518\n",
      "34152\n",
      "34153\n",
      "34154\n",
      "34155\n",
      "34156\n",
      "34157\n",
      "34158\n",
      "34159\n",
      "34463\n",
      "34417\n",
      "34467\n",
      "34136\n",
      "34137\n",
      "34134\n",
      "34135\n",
      "34644\n",
      "34133\n",
      "34408\n",
      "34409\n",
      "34406\n",
      "34138\n",
      "34402\n",
      "34400\n",
      "34401\n",
      "34529\n",
      "34722\n",
      "34484\n",
      "34720\n",
      "34480\n",
      "34481\n",
      "34520\n",
      "34283\n",
      "34522\n",
      "34304\n",
      "34172\n",
      "34449\n",
      "34086\n",
      "34717\n",
      "34382\n",
      "34423\n",
      "34314\n",
      "34558\n",
      "34207\n",
      "34490\n",
      "34102\n",
      "34438\n",
      "34433\n",
      "34432\n",
      "34437\n",
      "34435\n",
      "34434\n",
      "34637\n",
      "34636\n",
      "34630\n",
      "34633\n",
      "34532\n",
      "34530\n",
      "34531\n",
      "34537\n",
      "34534\n",
      "34274\n",
      "34036\n",
      "34035\n",
      "34034\n",
      "34033\n",
      "34032\n",
      "34031\n",
      "34030\n",
      "34395\n",
      "34039\n",
      "34506\n",
      "34117\n",
      "34349\n",
      "34348\n",
      "34114\n",
      "34347\n",
      "34220\n",
      "34110\n",
      "34058\n",
      "34509\n",
      "34113\n",
      "34626\n",
      "34217\n",
      "34620\n",
      "34591\n",
      "34508\n",
      "34592\n",
      "34629\n",
      "34351\n",
      "34593\n",
      "34221\n",
      "34193\n",
      "34212\n",
      "34338\n",
      "34189\n",
      "34232\n",
      "34025\n",
      "34026\n",
      "34027\n",
      "34028\n",
      "34029\n",
      "34332\n",
      "34180\n",
      "34187\n",
      "34186\n",
      "34185\n",
      "34184\n",
      "34169\n",
      "34168\n",
      "34161\n",
      "34160\n",
      "34163\n",
      "34162\n",
      "34350\n",
      "34166\n",
      "34194\n",
      "34195\n",
      "34196\n",
      "34197\n",
      "34191\n",
      "34192\n",
      "34425\n",
      "34198\n",
      "34199\n",
      "34661\n",
      "34424\n",
      "34223\n",
      "34426\n",
      "34427\n",
      "34421\n",
      "34224\n",
      "34588\n",
      "34587\n",
      "34586\n",
      "34584\n",
      "34277\n",
      "34581\n",
      "34580\n",
      "34178\n",
      "34179\n",
      "34284\n",
      "34422\n",
      "34286\n",
      "34363\n",
      "34173\n",
      "34170\n",
      "34171\n",
      "34176\n",
      "34177\n",
      "34174\n",
      "34175\n",
      "34561\n",
      "34609\n",
      "34471\n",
      "34569\n",
      "34183\n",
      "34469\n",
      "34577\n",
      "34330\n",
      "34582\n",
      "34451\n",
      "34450\n",
      "34453\n",
      "34452\n",
      "34455\n",
      "34454\n",
      "34219\n",
      "34218\n",
      "34590\n",
      "34216\n",
      "34215\n",
      "34214\n",
      "34594\n",
      "34595\n",
      "34596\n",
      "34210\n",
      "34069\n",
      "34213\n",
      "34066\n",
      "34576\n",
      "34288\n",
      "34574\n",
      "34575\n",
      "34570\n",
      "34571\n",
      "34077\n",
      "34674\n",
      "34579\n",
      "34132\n",
      "34130\n",
      "34638\n",
      "34208\n",
      "34209\n",
      "34188\n",
      "34150\n",
      "34200\n",
      "34201\n",
      "34202\n",
      "34203\n",
      "34204\n",
      "34205\n",
      "34206\n",
      "34371\n",
      "34360\n",
      "34211\n",
      "34268\n",
      "34269\n",
      "34266\n",
      "34267\n",
      "34264\n",
      "34265\n",
      "34260\n",
      "34705\n",
      "34704\n",
      "34707\n",
      "34701\n",
      "34151\n",
      "34703\n",
      "34182\n",
      "34662\n",
      "34368\n",
      "34139\n",
      "34385\n",
      "34181\n",
      "34387\n",
      "34386\n",
      "34723\n",
      "34474\n",
      "34528\n",
      "34473\n",
      "34147\n",
      "34146\n",
      "34145\n",
      "34143\n",
      "34142\n",
      "34372\n",
      "34140\n",
      "34477\n",
      "34476\n",
      "34475\n",
      "34470\n",
      "34378\n",
      "34472\n",
      "34149\n",
      "34148\n",
      "\n",
      "STAG will infer a species tree from each gene tree with all species present\n",
      "\n",
      "Test can run \"fastme -i /Volumes/McMinds/git_repos/gcmp_stan_symportal_2019/output/symbio_phylo/its2tree//../STAG_ResultsJun28/SimpleTest.phy -o /Volumes/McMinds/git_repos/gcmp_stan_symportal_2019/output/symbio_phylo/its2tree//../STAG_ResultsJun28/SimpleTest.tre\" - ok\n",
      "\n",
      "Processing gene trees:\n",
      "allgenes.pruned.treefile - Processed\n",
      "\n",
      "Examined 1 trees\n",
      "1 trees had all species present and will be used by STAG to infer the species tree\n",
      "\n",
      "STAG species tree: /Volumes/McMinds/git_repos/gcmp_stan_symportal_2019/output/symbio_phylo/STAG_ResultsJun28/SpeciesTree.tre\n",
      "\n",
      "\n",
      "Setting initial dates...\n",
      "Fitting in progress... get a first set of estimates\n",
      "         Penalised log-lik = -1355748 \n",
      "Optimising rates... dates... -1355748 \n",
      "Optimising rates... dates... -1355425 \n",
      "Optimising rates... dates... -1355235 \n",
      "Optimising rates... dates... -1355039 \n",
      "Optimising rates... dates... -1354828 \n",
      "Optimising rates... dates... -1354604 \n",
      "Optimising rates... dates... -1354398 \n",
      "Optimising rates... dates... -1354183 \n",
      "Optimising rates... dates... -1353899 \n",
      "Optimising rates... dates... -1353659 \n",
      "Optimising rates... dates... -1353471 \n",
      "Optimising rates... dates... -1353222 \n",
      "Optimising rates... dates... -1352971 \n",
      "Optimising rates... dates... -1352727 \n",
      "Optimising rates... dates... -1352469 \n",
      "Optimising rates... dates... -1352152 \n",
      "Optimising rates... dates... -1351940 \n",
      "Optimising rates... dates... -1351730 \n",
      "Optimising rates... dates... -1351459 \n",
      "Optimising rates... dates... -1351161 \n",
      "Optimising rates... dates... -1350974 \n",
      "Optimising rates... dates... -1350620 \n",
      "\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "!rm -rf {outdir + \"STAG_Results*\"}\n",
    "!python2 /Applications/STAG-1.0.0/stag/stag.py {profile_map_file} {gene_trees_folder}\n",
    "!cp -r {outdir + \"STAG_Results?*\"} {outdir + \"STAG_Results\"}\n",
    "!rm -rf {outdir + \"STAG_Results?*\"}\n",
    "profTree = skbio.TreeNode.read(outdir + \"STAG_Results/SpeciesTree.tre\").root_at_midpoint()\n",
    "profTree.write(file = outdir + \"STAG_Results/SpeciesTree_rooted.tree\")\n",
    "model = 'correlated'\n",
    "collapseMultis = True\n",
    "!Rscript {workdir + 'gcmp_stan_symportal/phylogenetics/chronogram.r'} {collapseMultis} {model} {outdir + \"STAG_Results/SpeciesTree_rooted.tree\"} {outdir + \"STAG_Results/SpeciesTree_\" + model + \"_chronos.tree\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "abunds = {}\n",
    "for string,profdef,names in zip(df.loc['Average defining sequence proportions and [stdev]'], df.loc['ITS2 type profile'], df.columns):\n",
    "    abunds[names] = {}\n",
    "    for substr,div in zip(string.split('-'),profdef.replace('/','-').split('-')):\n",
    "        abunds[names][div] = float(re.sub('\\[.*\\]','',substr)) * 1000 # need integer values for some reason! provided decimal values were at precision of 1e-3\n",
    "abunds = pd.DataFrame.from_dict(abunds, orient='index')\n",
    "abunds[pd.isna(abunds)] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Volumes/McMinds/git_repos/gcmp_stan_symportal_2019/output/symbio_phylo/allgenes_rooted.treefile'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "divTree = skbio.TreeNode.read(outdir + \"allgenes.treefile\").root_at_midpoint()\n",
    "divTree.write(file = outdir + \"allgenes_rooted.treefile\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqdf = pd.read_csv(outdir + \"allgenes.uniqueseq.phy\", sep=\"\\s+\", index_col=0, header=None)\n",
    "f = open(outdir + \"allgenes.uniqueseq.replicated.phy\", 'w')\n",
    "f.write(str(len(alldivs)) + ' ')\n",
    "with open(outdir + \"allgenes.uniqueseq.phy\") as f2:\n",
    "    f.write(next(f2).split(' ')[1])\n",
    "for div in allnames.keys():\n",
    "    for name in allnames[div]:\n",
    "        newname = name.split('_')[0]\n",
    "        newname = newname + ' ' * (10 - len(newname))\n",
    "        f.write(newname + list(seqdf.loc[div])[0] + '\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!iqtree -mfreq F,FO -mrate E,I,G,I+G,R,H,I+H -s {outdir + \"its2_all_clades_aligned.fasta\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ete3 import Tree\n",
    "gene_trees_folder = outdir + \"its2tree/\"\n",
    "pruned_tree = gene_trees_folder + \"its2_all_clades_aligned.fasta.pruned.treefile\"\n",
    "!mkdir {gene_trees_folder}\n",
    "!sed 's/_seed_//g' {outdir + \"its2_all_clades_aligned.fasta.treefile\"} > {pruned_tree}\n",
    "tree = Tree(pruned_tree)\n",
    "tree.prune(alldivs, preserve_branch_length=True)\n",
    "tree.write(format=1, outfile=pruned_tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in (\"28S\",\"23S\",\"cob\",\"coi\",\"elf\",\"psba\"):\n",
    "    !iqtree -pre {outdir + \"forindividual_tree_\"+gene+\"_\"} -mtree -mrate E,I,G,I+G,R,H,I+H -mfreq F,FO -s {workdir + \"raw_data/symbio_phylo/\"+gene+\"/forindividual\\ tree\\ \"+gene+\".fas\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ete3\n",
    "leaves = []\n",
    "for leaf in tree:\n",
    "    leaves.append(leaf.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for prof in profs.keys():\n",
    "    divs = profs[prof].replace('/',',').replace('-',',') \n",
    "    if not set(divs.split(',')).issubset(leaves):\n",
    "        print(profs[prof])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "treeS = Tree(\"/Volumes/McMinds/git_repos/gcmp_stan_symportal_2019/output/symbio_phylo/STAG_ResultsJun26/SpeciesTree.tree\")\n",
    "newleaves = []\n",
    "for leaf in treeS:\n",
    "    newleaves.append(leaf.name)\n",
    "for prof in profs.keys():\n",
    "    if prof not in newleaves:\n",
    "        print(profs[prof])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "treeS = Tree(\"/Volumes/McMinds/git_repos/gcmp_stan_symportal_2019/output/symbio_phylo/STAG_ResultsJun26/SpeciesTree.tre\")\n",
    "newleaves = []\n",
    "for leaf in treeS:\n",
    "    newleaves.append(leaf.name)\n",
    "for prof in profs.keys():\n",
    "    if prof in newleaves:\n",
    "        print(profs[prof])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
